lens/phrase.ml

```diff
@@ -122,61 +123,96 @@ let[@inline always] val_numeric op_i op_f v1 v2 =
+             arg) )
+
 let rec eval expr get_val =
   let open Value in
   match expr with
   | Constant c -> c
   | Var v -> (
-    try get_val v with Not_found ->
-      failwith ("Could not find column " ^ v ^ ".") )
-  | InfixAppl (op, a1, a2) -> (
+    try get_val v
+    with Not_found -> failwith ("Could not find column " ^ v ^ ".") )
+  | InfixAppl (op, a1, a2) ->
       let a1 = eval a1 get_val in
       let a2 = eval a2 get_val in
-      match op with
-      | Binary.GreaterEqual ->
-          val_typ_cmp ( >= ) ( >= ) ( >= ) ( >= ) ( >= ) a1 a2
-      | Binary.Greater -> val_typ_cmp ( > ) ( > ) ( > ) ( > ) ( > ) a1 a2
-      | Binary.LessEqual ->
-          val_typ_cmp ( <= ) ( <= ) ( <= ) ( <= ) ( <= ) a1 a2
-      | Binary.Less -> val_typ_cmp ( < ) ( < ) ( < ) ( < ) ( < ) a1 a2
-      | Binary.Equal -> val_typ_cmp ( = ) ( = ) ( = ) ( = ) ( = ) a1 a2
-      | Binary.LogicalAnd -> box_bool (unbox_bool a1 && unbox_bool a2)
-      | Binary.LogicalOr -> box_bool (unbox_bool a1 || unbox_bool a2)
-      | Binary.Plus -> val_numeric ( + ) ( +. ) a1 a2
-      | Binary.Multiply ->
-          val_numeric (fun v v' -> v * v') (fun v v' -> v *. v') a1 a2
-      | Binary.Minus -> val_numeric ( - ) ( -. ) a1 a2
-      | Binary.Divide -> val_numeric ( / ) ( /. ) a1 a2 )
+      eval_infix op a1 a2
   | TupleLit l -> eval (List.hd l) get_val
-  | UnaryAppl (op, arg) -> (
-    match op with
-    | Unary.Not ->
-        let res = eval arg get_val in
-        let res = not (unbox_bool res) in
-        box_bool res
-    | Unary.Minus -> (
-        let res = eval arg get_val in
-        match res with
-        | Value.Float v -> -.v |> box_float
-        | Value.Int v -> -v |> box_int
-        | _ ->
-            failwith
-              (Format.asprintf
-                 "Value '%a' does not support the unary minus operator."
-                 Value.pp res) ) )
+  | UnaryAppl (op, arg) ->
+      let res = eval arg get_val in
+      eval_unary op res
   | In (names, vals) ->
       let find = List.map ~f:get_val names in
       let equal s t = List.for_all2_exn ~f:Value.equal s t in
       let res = List.mem ~equal vals find in
       box_bool res
   | Case (inp, cases, otherwise) -> (
       let inp =
-        match inp with None -> Value.Bool true | Some inp -> eval inp get_val
+        match inp with
+        | None -> Value.Bool true
+        | Some inp -> eval inp get_val
       in
       match List.find ~f:(fun (k, _v) -> eval k get_val = inp) cases with
       | Some (_, v) -> eval v get_val
       | None -> eval otherwise get_val )
 
+let rec partial_eval expr ~lookup =
+  match expr with
+  | Constant c -> Constant c
+  | Var v ->
+      lookup v
+      |> Option.map ~f:(fun v -> Constant v)
+      |> Option.value ~default:(Var v)
+  | InfixAppl (op, a1, a2) -> (
+      let a1 = partial_eval a1 ~lookup in
+      let a2 = partial_eval a2 ~lookup in
+      match (op, a1, a2) with
```

Comment:
Consider `begin...end` instead of brackets again (it took me a while to spot the brackets and I was confused). Also, the spacing seems very inconsistent below

Reply:
>--indicate-nested-or-patterns={space|unsafe-no}
           Control whether or not to indicate nested or-pattern using indentation. space starts lines of nested or-patterns with " |" rather than "| ".
           unsafe-no does not indicate nested or-patterns. Warning: this can produce confusing code where a short body of a match case is visually hidden
           by surrounding long patterns, leading to misassociation between patterns and body expressions. The default value is space.

The "inconsistent spacing" is an indication of nested or patterns, see manual extract above. I didn't want to set it to `unsafe-no`. If the match was missing a bracket it would be indented differently.